src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/Renamer.cs

```diff
@@ -78,18 +83,131 @@ public async void OnWorkspaceChangedAsync(object sender, WorkspaceChangeEventArg
         public async Task RenameAsync(Project project)
         {
             bool isCaseSensitive = await IsCompilationCaseSensitiveAsync(project);
-            IRenameStrategy renameStrategy = GetStrategy(isCaseSensitive);
-            if (renameStrategy != null)
-                await renameStrategy.RenameAsync(project, _oldFilePath, _newFilePath, isCaseSensitive);
+            await RenameAsync(project, _oldFilePath, _newFilePath, isCaseSensitive);
         }
 
-        private IRenameStrategy GetStrategy(bool isCaseSensitive)
+        private async Task RenameAsync(Project project, string oldFilePath, string newFilePath, bool isCaseSensitive)
         {
-            var strategies = new IRenameStrategy[] {
-                new SimpleRenameStrategy(_threadingService, _userNotificationServices, _environmentOptions, _roslynServices)
-            };
+            string oldNameBase = Path.GetFileNameWithoutExtension(oldFilePath);
 
-            return strategies.FirstOrDefault(s => s.CanHandleRename(_oldFilePath, _newFilePath, isCaseSensitive));
+            if (!await ShouldRenameAsync(project, oldNameBase, newFilePath, isCaseSensitive))
+                return;
+
+            bool userConfirmed = await CheckUserConfirmation(oldNameBase);
+            if (!userConfirmed)
+                return;
+
+            Solution renamedSolution = await GetRenamedSolutionAsync(project, oldNameBase, newFilePath, isCaseSensitive);
+            if (renamedSolution == null)
+                return;
+
+            await _threadingService.SwitchToUIThread();
+            bool renamedSolutionApplied = _roslynServices.ApplyChangesToSolution(project.Solution.Workspace, renamedSolution);
+
+            if (!renamedSolutionApplied)
+            {
+                string failureMessage = string.Format(CultureInfo.CurrentCulture, VSResources.RenameSymbolFailed, oldNameBase);
+                await _threadingService.SwitchToUIThread();
```

Comment:
Is this redundant?

Reply:
No, tests will fail if this is not present

Reply:
I can try and find a way to unify this logic though